mantidproject/mantid
# Add copy paste for shapes in the instrument viewer

**Description of work.**

When drawing shapes in the instrument viewer, it is now possible to copy and paste the selected ones using Ctrl+C / Ctrl+V . They are pasted with an offset from their origin, and the offset increases at each past so they are not stacked if pasted multiple times.

The exception are the sectors, which are pasted with an angular offset, so they form a bigger sector.



**To test:**
- Open the instrument viewer and check, for each shape (including the freely drawn shape), they are correctly copied and pasted, and they behave in an intuitive way.



WIP for  #26906 . 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
